drivers/hmis/app/models/hmis/workflow_execution/engine.rb

```diff
@@ -0,0 +1,151 @@
+require 'dentaku'
+
+# Manages the execution of workflow instances, handling state transitions,
+# task assignments, and message processing according to the workflow template.
+module Hmis::WorkflowExecution
+  class Engine
+    attr_reader :template, :instance, :message_handler, :assignment_handler
+
+    def initialize(workflow_instance, message_handler:, assignment_handler:)
+      @instance = workflow_instance
+      @template = workflow_instance.template
+      @message_handler = message_handler
+      @assignment_handler = assignment_handler
+    end
+
+    def active_steps
+      instance.steps.where(status: ['available', 'in_progress'])
+    end
+
+    def start_workflow!(user:)
+      template.nodes.entrypoints.each do |node|
+        visit_node(node)
+      end
+      log_event('start_workflow', user: user)
+    end
+
+    def start_step!(step, user:)
+      step.start!
+      process_triggers(step.node, 'start_step')
+      log_event('start_step', user: user, step: step)
+    end
+
+    def complete_step!(step, user:, submitted_values:)
+      step.submitted_values = submitted_values
+      step.complete!
```

Comment:
Is `complete!` implemented? I don't see it but maybe there is rails magic with the statuses?

Reply:
It's magic from the `aasm` gem. Methods correspond to the events in the aasm config for this model:
```ruby
 # step.rb
 aasm column: 'status' do
   # status...
   event :complete do # defines `complete`!
     # ...
   end
 end
```

Reply:
I see, thanks